@navikt/core/react/src/util/hooks/useScrollLock.ts

```diff
@@ -0,0 +1,291 @@
+  win.addEventListener("resize", handleResize);
+
+  return () => {
+    if (resizeRaf) {
+      cancelAnimationFrame(resizeRaf);
+    }
+    cleanup();
+    win.removeEventListener("resize", handleResize);
+  };
+}
+
+class ScrollLocker {
+  lockCount = 0;
+  restore: (() => void) | null = null;
+  timeoutLock = Timeout.create();
+  timeoutUnlock = Timeout.create();
+
+  /**
+   * Aquires a new lock
+   * - If first lock, lock document-scroll.
+   * - If not first lock, do nothing.
+   */
+  acquire(referenceElement: Element | null) {
+    this.lockCount += 1;
+    if (this.lockCount === 1 && this.restore === null) {
+      this.timeoutLock.start(0, () => this.lock(referenceElement));
+    }
+    return this.release;
+  }
+
+  /**
+   * Releases a lock
+   * - If last lock, unlock document-scroll.
+   * - If not last lock, do nothing.
+   */
+  release = () => {
+    this.lockCount -= 1;
+    if (this.lockCount === 0 && this.restore) {
+      this.timeoutUnlock.start(0, this.unlock);
+    }
+  };
+
+  private unlock = () => {
+    if (this.lockCount === 0 && this.restore) {
+      this.restore?.();
+      this.restore = null;
+    }
+  };
+
+  private lock(referenceElement: Element | null) {
+    if (this.lockCount === 0 || this.restore !== null) {
+      return;
+    }
+
+    const doc = ownerDocument(referenceElement);
+    const html = doc.documentElement;
+    const htmlOverflowY = ownerWindow(html).getComputedStyle(html).overflowY;
+
+    /* If the site author already hid overflow on <html>, respect it and bail out. */
+    if (htmlOverflowY === "hidden" || htmlOverflowY === "clip") {
+      this.restore = NOOP;
+      return;
+    }
+
+    const isOverflowHiddenLock = isIOS || !hasInsetScrollbars(referenceElement);
+
+    /**
+     * On iOS, scroll locking does not work if the navbar is collapsed.
+     * Due to side-effects and bugs that arise on iOS, it must be researched extensively before
+     * being enabled to ensure it doesn't cause the following issues:
+     * - Textboxes must scroll into view when focused, nor cause a glitchy scroll animation.
+     * - The navbar must not force itself into view and cause layout shift.
+     * - Scroll containers must not flicker upon closing a popup when it has an exit animation.
+     */
+    this.restore = isOverflowHiddenLock
+      ? preventScrollBasic(referenceElement)
+      : preventScrollStandard(referenceElement);
+  }
+}
+
+const SCROLL_LOCKER = new ScrollLocker();
+
+/**
+ * Locks the scroll of the document when enabled.
+ * @param enabled - Whether to enable the scroll lock.
+ */
+function useScrollLock(params: {
+  enabled: boolean;
+  mounted: boolean;
+  open: boolean;
+  referenceElement?: Element | null;
+}) {
+  const { enabled = true, mounted, open, referenceElement = null } = params;
+
+  /**
+   * https://github.com/mui/base-ui/issues/1135
+   * This only applies changes in the state where element is:
+   * - No longer open
+   * - Is mounted (and probably animating out)
```

Comment:
The line "This only applies..." is a bit confusing to me. It's not obvious what "changes" and "state" refers to. Maybe what's most relevant to explain is that !open + mounted probably means we are animating out?

```suggestion
   * mounted + not open = probably animating out
```
Might be better to put that right after the `if` in question though.